dashboard/src/components/Breadcrumb/TreeBreadcrumb.tsx

```diff
@@ -22,7 +22,43 @@ const TreeBreadcrumb = ({
   searchParams: LinkProps['search'];
   locationMessage: MessageDescriptor['id'];
 }): JSX.Element => {
-  const treeId = useRouterState({ select: s => s.location.state.id });
+  const historyState = useRouterState({
+    select: s => s.location.state,
+  });
+
+  const { treeName, branch, id: treeId } = historyState;
+  const canGoDirect = treeName && branch && treeId;
+
+  const treeParams: {
+    to: '/tree/$treeName/$branch/$hash' | '/tree/$treeId';
+    params:
+      | { treeName?: string; branch?: string; hash?: string }
+      | { treeId?: string };
+  } = useMemo(
+    () =>
+      canGoDirect
+        ? {
+            to: '/tree/$treeName/$branch/$hash',
+            params: { treeName: treeName, branch: branch, hash: treeId },
+          }
+        : { to: '/tree/$treeId', params: { treeId: treeId } },
+    [canGoDirect, treeName, branch, treeId],
+  );
```

Comment:
You don't need to type the `treeParams` if you add `as const` to the objects

```suggestion
  const treeParams = useMemo(
    () =>
      canGoDirect
        ? {
            to: '/tree/$treeName/$branch/$hash',
            params: { treeName: treeName, branch: branch, hash: treeId },
          } as const
        : { to: '/tree/$treeId', params: { treeId: treeId } } as const,
    [canGoDirect, treeName, branch, treeId],
  );
```